lib/report_portal/cucumber/parallel_report.rb

```diff
@@ -35,65 +35,65 @@ def initialize(desired_time)
         ReportPortal.last_used_time = 0
         set_parallel_tests_vars
         if ParallelTests.first_process?
-          File.open(file_with_launch_id, 'w') do |f|
-            f.flock(File::LOCK_EX)
-            start_launch(desired_time, @cmd_args_of_parallel_tests)
-            f.write(ReportPortal.launch_id)
-            f.flush
-            f.flock(File::LOCK_UN)
-          end
-          @folder_creation_tracking_file = Pathname(Dir.tmpdir)  + "folder_creation_tracking_#{ReportPortal.launch_id}.lck"
-          File.open(@folder_creation_tracking_file, 'w+') do |f|
-            f.flock(File::LOCK_EX)
-            f.flush
-            f.flock(File::LOCK_UN)
-          end
+          start_launch(desired_time)
         else
           start_time = monotonic_time
           loop do
-            break if File.exist?(file_with_launch_id)
+            break if File.exist?(lock_file)
             if monotonic_time - start_time > wait_time_for_launch_start
               raise "File with launch ID wasn't created after waiting #{wait_time_for_launch_start} seconds"
             end
             sleep 0.5
           end
-          File.open(file_with_launch_id, 'r') do |f|
-            f.flock(File::LOCK_SH)
```

Comment:
Could you please explain why this line was removed?
My understanding is that without it `File#read` may read the file before lock for writing the file is removed.

Reply:
each cucumber is a separate process,  LOCK_EX  will do the job 

https://github.com/reportportal/agent-ruby/blob/e62541eb7242e3e454c029161ad7d28d2a89bebd/lib/report_portal/cucumber/report.rb#L96

https://ruby-doc.org/core-2.0.0/File.html#method-i-flock
> LOCK_EX   | Exclusive lock. Only one process may hold an
>           | exclusive lock for a given file at a time.

Reply:
The locks are advisory. If the program doesn't check if the file is locked, it can do anything even if the file is locked exclusively.
That's why when writing the file, exclusive (write) lock is set.
And when reading the file, shared (read) lock is set. That will prevent usage of the file before the exclusive lock is released.
https://stackoverflow.com/a/20452541
https://linux.die.net/man/2/flock

> flock() places advisory locks only; given suitable permissions on a file, a process is free to ignore the use of flock() and perform I/O on the file.

Reply:
but we use `flock(File::LOCK_EX)`  when writing to file not `flock()` check the second response in StackOverflow that you have linked.

Reply:
http://ruby-doc.org/core-2.5.3/File.html#method-i-flock
Please take a look at this example from Ruby documentation. It uses LOCK_EX when writing the file and LOCK_SH when reading it. If there were no LOCK_SH when reading, it could have read an empty or incompletely written file.

And in this example (https://github.com/ruby/ruby/blob/d48783bb0236db505fe1205d1d9822309de53a36/lib/pstore.rb#L379) from Ruby code the same happens. When the file is written, there is LOCK_EX flock, and it's read there's LOCK_SH flock.

Reply:
this would be needed if tests were triggered in treads and not as separate processes. Do you have a use case where formatter would be used in a multi-treaded run?

Reply:
`flock` sets locks that can be seen system-wide. If one process sets a lock, then another process can check for the presence of it.

```
Process.fork do
  File.open("counter", File::RDWR|File::CREAT, 0644) do |f|
    f.flock(File::LOCK_EX)
    sleep 4
  end
end

Process.fork do
  File.open("counter", "r") do |f|
    puts Time.now
    f.flock(File::LOCK_SH)
    puts Time.now
  end
end

Process.wait
```
As you can see the second process waits for 4 seconds.

Reply:
ok, I got your point, you are correct, without lock nothing stops you from reading the file. will revert the change.